If you are running a laptop, NVDA will now announce changes to whether a charger is plugged in and more. #4612

derekriemer opened the pull request:
If you are running a laptop, NVDA will now announce changes to whether a charger is plugged in.
Changes to the orientation of the screen are now announced, For example, turning a laptop so it is situated with the longer side horizontal, you'll hear or feel "Landscape"

Fixes #4574. Fixes #4612.

Commit:
If you are running a laptop, NVDA will now announce changes to whether a charger is plugged in.
Changes to the orientation of the screen are now announced, For example, turning a laptop so it is situated with the longer side horizontal, you'll hear or feel "Landscape"

Commit:
Add copyright info

Commit:
Fix gramar error

jcsteh commented on source/core.py:
```diff
@@ -250,6 +250,48 @@ def onEndSession(evt):
 	import windowUtils
 	class MessageWindow(windowUtils.CustomWindow):
 		className = u"wxWindowClassNR"
+		batteryCache = None
```
Is there a more descriptive name that more clearly states what this stores; e.g. lastBatteryStatus or similar?

jcsteh commented on source/core.py:
```diff
@@ -250,6 +250,48 @@ def onEndSession(evt):
 	class MessageWindow(windowUtils.CustomWindow):
 		className = u"wxWindowClassNR"
+		batteryCache = None
+		REASON_INITIALIZE = 1
```
I don't think you need this constant. The function could instead accept a bool parameter; e.g. `init=True`.

jcsteh commented on source/core.py:
```diff
@@ -250,6 +250,48 @@ def onEndSession(evt):
+			#Just define these constants here, so we don't have to import win32con
+			WM_POWERBROADCAST = 0x218
+			WM_DISPLAYCHANGE = 0x7e
+			PBT_APMPOWERSTATUSCHANGE = 0xA
```
Please define these at class level so they don't have to be redefined each time the function runs.

jcsteh commented on source/core.py:
```diff
@@ -250,6 +250,48 @@ def onEndSession(evt):
+				self.say_battery_status()
+			# Resolution detection comes from an article found at https://msdn.microsoft.com/en-us/library/ms812142.aspx.
+			elif msg == WM_DISPLAYCHANGE:
+				ui.message("Landscape" if lParam%0x10000 > lParam/0x10000 else "Portrait")
```
These strings need to be made translatable, including translator comments. That requires that both strings are on separate lines so the translator comment can apply to the correct string.

jcsteh commented on source/core.py:
```diff
@@ -250,6 +250,48 @@ def onEndSession(evt):
+			elif msg == WM_DISPLAYCHANGE:
+				ui.message("Landscape" if lParam%0x10000 > lParam/0x10000 else "Portrait")
+
+
```
Extraneous white space.

jcsteh commented on source/core.py:
```diff
@@ -250,6 +250,48 @@ def onEndSession(evt):
+				ui.message("Landscape" if lParam%0x10000 > lParam/0x10000 else "Portrait")
+
+
+		def say_battery_status(self, reason=None):
```
1. Please rename to `handlePowerStatusChange` or similar.
2. Do you need the initialisation parameter at all? Could you just check whether the cached value has been set yet?

jcsteh commented on source/core.py:
```diff
@@ -250,6 +250,48 @@ def onEndSession(evt):
+			import winKernel
+			UNKNOWN_BATTERY_STATUS = 0xFF
+			AC_ONLINE = 0X1
+			NO_SYSTEM_BATTERY = 0X80
```
Please move constants to class level.

jcsteh commented on source/core.py:
```diff
@@ -250,6 +250,48 @@ def onEndSession(evt):
+				return
+			if sps.BatteryFlag & NO_SYSTEM_BATTERY or sps.ACLineStatus == self.batteryCache:
+				return
+			#Necessary because sometimes, this event double fires, and also, if the battery decreases by 3 percent, this event occurs.
```
Do you mean "decreases to 3 percent", rather than "by 3 percent"?

derekriemer replied:
Nope.


On 9/4/2016 11:51 PM, James Teh wrote:
>
> In source/core.py 
> <https://github.com/nvaccess/nvda/pull/6320#discussion_r77472714>:
>
> > +				ui.message("Landscape" if lParam%0x10000 > lParam/0x10000 else "Portrait")
> > +
> > +
> > +		def say_battery_status(self, reason=None):
> > +			#Mostly taken from script_say_battery_status, but modified.
> > +			import ui
> > +			import winKernel
> > +			UNKNOWN_BATTERY_STATUS = 0xFF
> > +			AC_ONLINE = 0X1
> > +			NO_SYSTEM_BATTERY = 0X80
> > +			sps = winKernel.SYSTEM_POWER_STATUS()
> > +			if not winKernel.GetSystemPowerStatus(sps) or sps.BatteryFlag is UNKNOWN_BATTERY_STATUS:
> > +				return
> > +			if sps.BatteryFlag & NO_SYSTEM_BATTERY or sps.ACLineStatus == self.batteryCache:
> > +				return
> > +			#Necessary because sometimes, this event double fires, and also, if the battery decreases by 3 percent, this event occurs.
>
> Do you mean "decreases to 3 percent", rather than "by 3 percent"?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/nvaccess/nvda/pull/6320/files/69b19519a04ca0f5c647bb9023d1eb3d8c9f5e47#r77472714>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AFGivQT70FERf51UY9dEUcT6uDFfhItIks5qm63_gaJpZM4JxF3F>.
>

-- 
------------------------------------------------------------------------


    Derek Riemer

  * Department of computer science, third year undergraduate student.
  * Proud user of the NVDA screen reader.
  * Open source enthusiast.
  * Member of Bridge Cu
  * Avid skiier.

Websites:
Honors portfolio <http://derekriemer.com>
Awesome little hand built weather app! 
<http://django.derekriemer.com/weather/>

email me at derek.riemer@colorado.edu <mailto:derek.riemer@colorado.edu>
Phone: (303) 906-2194

jcsteh replied:
So this seriously fires for every 3% the battery decreases; i.e. 33 times per charge cycle? That seems very odd.

derekriemer replied:
I found MSDN documentation that stated the following.

"This event can occur when battery life drops to less than 5 minutes, or 
when the percentage of battery life drops below 10 percent, or if the 
battery life changes by 3 percent.

"

https://msdn.microsoft.com/en-us/library/windows/desktop/aa372715(v=vs.85).aspx


On 9/5/2016 2:01 AM, James Teh wrote:
>
> In source/core.py 
> <https://github.com/nvaccess/nvda/pull/6320#discussion_r77482716>:
>
> > +				ui.message("Landscape" if lParam%0x10000 > lParam/0x10000 else "Portrait")
> > +
> > +
> > +		def say_battery_status(self, reason=None):
> > +			#Mostly taken from script_say_battery_status, but modified.
> > +			import ui
> > +			import winKernel
> > +			UNKNOWN_BATTERY_STATUS = 0xFF
> > +			AC_ONLINE = 0X1
> > +			NO_SYSTEM_BATTERY = 0X80
> > +			sps = winKernel.SYSTEM_POWER_STATUS()
> > +			if not winKernel.GetSystemPowerStatus(sps) or sps.BatteryFlag is UNKNOWN_BATTERY_STATUS:
> > +				return
> > +			if sps.BatteryFlag & NO_SYSTEM_BATTERY or sps.ACLineStatus == self.batteryCache:
> > +				return
> > +			#Necessary because sometimes, this event double fires, and also, if the battery decreases by 3 percent, this event occurs.
>
> So this seriously fires for every 3% the battery decreases; i.e. 33 
> times per charge cycle? That seems very odd.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/nvaccess/nvda/pull/6320/files/69b19519a04ca0f5c647bb9023d1eb3d8c9f5e47#r77482716>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AFGivTPAFiBRlrwWAKJXUIEKPN6Vkn7Nks5qm8x2gaJpZM4JxF3F>.
>

-- 
------------------------------------------------------------------------


    Derek Riemer

  * Department of computer science, third year undergraduate student.
  * Proud user of the NVDA screen reader.
  * Open source enthusiast.
  * Member of Bridge Cu
  * Avid skiier.

Websites:
Honors portfolio <http://derekriemer.com>
Awesome little hand built weather app! 
<http://django.derekriemer.com/weather/>

email me at derek.riemer@colorado.edu <mailto:derek.riemer@colorado.edu>
Phone: (303) 906-2194

jcsteh replied:
Fair enough. I only asked because I remembered you mentioning critical battery on IRC, but not the 3% thing. Thanks for clarifying. Carry on. :)

jcsteh commented on source/core.py:
```diff
@@ -250,6 +250,48 @@ def onEndSession(evt):
+				text = _("Plugged in. Batteri is ")
+			else:
+				text = _("Not plugged in. Battery is ")
+			text += _("%d percent full") % sps.BatteryLifePercent
```
1. It's best not to split translatable strings in the middle of a phrase like this, as different languages have different rules about how stuff is ordered.
2. I think this could be briefer; e.g. "Plugged in, battery 100%".
3. Given both of the above, strings like `_("Plugged in, battery %d percent")` are probably best. You might be able to get away with `_("Plugged in, ")` and `_("battery %d percent")`, but since you need tow strings anyway, it's probably just easier to make the whole thing separate for the two cases.
4. These strings need translator comments.

josephsl commented:
Hi,

Issues #4574 and #4612 combined into a single PR package. Thanks.

Commit:
Address review actions

derekriemer commented:
@jcsteh This is ready for another review.

derekriemer commented:
@jcsteh This is ready for another review, when time allows.

derekriemer commented:
Hi:
Just wanted to ask kindly for a review again when time allows.
Thanks.

jcsteh commented on source/core.py:
```diff
@@ -251,6 +251,54 @@ def onEndSession(evt):
 	class MessageWindow(windowUtils.CustomWindow):
 		className = u"wxWindowClassNR"
+		oldBatteryStatus = None
+		hasBeenInitialized = False
```
I still wonder whether we can get rid of this variable and instead just have an argument (onlyCache=True or similar) for handlePowerStatusChange which only gets passed in the constructor.

jcsteh commented on source/core.py:
```diff
@@ -251,6 +251,54 @@ def onEndSession(evt):
+			import ui
+			if msg == self.WM_POWERBROADCAST and wParam == self.PBT_APMPOWERSTATUSCHANGE:
+				self.handlePowerStatusChange()
+			# Resolution detection comes from an article found at https://msdn.microsoft.com/en-us/library/ms812142.aspx.
```
I think it'd be clearer if this comment was inside the `elif` block.

jcsteh commented on source/core.py:
```diff
@@ -251,6 +251,54 @@ def onEndSession(evt):
+				self.handlePowerStatusChange()
+			# Resolution detection comes from an article found at https://msdn.microsoft.com/en-us/library/ms812142.aspx.
+			elif msg == self.WM_DISPLAYCHANGE:
+				if lParam%0x10000 > lParam/0x10000:
```
This would be more readable using the winUser.HIWORD and winUser.LOWORD functions, since this is a parameter composed of two words and you want to examine them separately.

jcsteh commented on source/core.py:
```diff
@@ -251,6 +251,54 @@ def onEndSession(evt):
+				return
+			if sps.BatteryFlag & self.NO_SYSTEM_BATTERY or sps.ACLineStatus == self.oldBatteryStatus:
+				return
+			#Necessary because sometimes this event double fires, and  if the battery decreases by 3 percent, this event occurs.
```
To make this clearer, I think it'd be good to move this comment up a bit and be explicit about what is necessary; i.e. caching the previous battery status.

jcsteh commented on source/core.py:
```diff
@@ -251,6 +251,54 @@ def onEndSession(evt):
+			MessageWindow.oldBatteryStatus = sps.ACLineStatus
+			if MessageWindow.hasBeenInitialized == False:
+				MessageWindow.hasBeenInitialized = True
+				return #We don't actually want to output anything, just initialize the cache.
```
See above regarding an onlyCache argument. Also, `if not blah` is more readable for a boolean.

derekriemer replied:
Haha, I was working on lots of c++ at the time, so my brain was in type safety mode.

jcsteh reviewed (changes requested):
Thanks for your work. Sorry it took me so long to get back to this.

Commit:
Remove the initializing cache, and add it as an argument.

Commit:
Move comment into block

Commit:
Use hiword and lowword

Commit:
Move the comment up

derekriemer commented:
@jcsteh Ready for another round.

jcsteh commented on source/core.py:
```diff
@@ -1,7 +1,7 @@
 #core.py
 #A part of NonVisual Desktop Access (NVDA)
-#Copyright (C) 2006-2016 NV Access Limited, Aleksey Sadovoy, Christopher Toth, Joseph Lee, Peter Vágner
+#Copyright (C) 2006-2016 NV Access Limited, Aleksey Sadovoy, Christopher Toth, Joseph Lee, Peter Vágner, Derek Riemer
```
2016 should be 2017.

jcsteh commented on source/core.py:
```diff
@@ -262,6 +262,55 @@ def onEndSession(evt):
 	import windowUtils
 	class MessageWindow(windowUtils.CustomWindow):
 		className = u"wxWindowClassNR"
+		oldBatteryStatus = None
```
Any reason you put this on the class and then set it on the class later? It seems to me this would be more appropriate as an instance variable.

jcsteh commented on source/core.py:
```diff
@@ -262,6 +262,55 @@ def onEndSession(evt):
 	class MessageWindow(windowUtils.CustomWindow):
 		className = u"wxWindowClassNR"
+		oldBatteryStatus = None
+		hasBeenInitialized = False
```
No longer necessary.

derekriemer replied:
lol, I was so excited to remove this I forgot to.

jcsteh commented on source/core.py:
```diff
@@ -262,6 +262,55 @@ def onEndSession(evt):
+			import winUser
+			if msg == self.WM_POWERBROADCAST and wParam == self.PBT_APMPOWERSTATUSCHANGE:
+				self.handlePowerStatusChange()
+			elif msg == self.WM_DISPLAYCHANGE:
```
I normally wouldn't suggest this, but because this is in the middle of a very general window procedure which will probably get extended to support all sorts of things, I'd prefer the code inside this block to handle orientation was split into its own method too.

jcsteh commented on source/core.py:
```diff
@@ -262,6 +262,55 @@ def onEndSession(evt):
+			#This cache is Necessary because sometimes this event double fires, and  if the battery decreases by 3 percent, this event occurs.
+			if sps.BatteryFlag & self.NO_SYSTEM_BATTERY or sps.ACLineStatus == self.oldBatteryStatus:
+				return
+			MessageWindow.oldBatteryStatus = sps.ACLineStatus
```
See above regarding this being an instance variable.

jcsteh reviewed (changes requested):
Nearly there!

Commit:
Review actions, change from using an init param to using a try/except/finally

derekriemer commented:
@jcsteh fixes applied

jcsteh commented on source/core.py:
```diff
@@ -274,36 +272,41 @@ class MessageWindow(windowUtils.CustomWindow):
+				#This cache is Necessary because sometimes this event double fires, and  if the battery decreases by 3 percent, this event occurs.
+				if sps.BatteryFlag & self.NO_SYSTEM_BATTERY or sps.ACLineStatus == self.oldBatteryStatus:
+					return
+			except AttributeError:
```
This pattern for checking init is fine, except I'd prefer having the cache initialized to None in the constructor and then testing that here, rather than relying on AttributeError. It just looks cleaner because you know what the initialized state is meant to be. It also avoids hiding a typo if that got changed by mistake in future. :)

jcsteh reviewed (changes requested):
Sorry, one more. :(

Commit:
fix this last review action
I took a longer but more clear approach.

derekriemer commented:
@jcsteh I changed this. It should be ready for review again. TA

jcsteh commented on source/core.py:
```diff
@@ -262,6 +262,62 @@ def onEndSession(evt):
+			sps = winKernel.SYSTEM_POWER_STATUS()
+			if not winKernel.GetSystemPowerStatus(sps) or sps.BatteryFlag is self.UNKNOWN_BATTERY_STATUS:
+				return
+			#This cache is Necessary because sometimes this event double fires, and  if the battery decreases by 3 percent, this event occurs.
```
nit: I don't think this comment is needed any more, as it's explained better later. Certainly, if you do want to keep it, it should be moved closer to the cache code.

Commit:
Fix NIT

derekriemer commented:
Fix NIT 5930c5732

leonardder commented:
It seems that the screen orientation reporting also reports a screen orientation change in cases when it shouldn't. For example, When i change the display scaling in the settings on a pc with the Windows 10 creators update, NVDA sometimes says landscape before reading the new scaling value.

Would it be possible to cache the display orientation so that changes from landscape to landscape will be suppressed?

derekriemer commented:
That probably does a broadcast, and change from portrate to landscape are when it goes from taller than wide to wider than tall. I can cache, but It'll be like (w,h), and if either of (w,h) change, don't speak.

jcsteh commented:
I'd be inclined to cache the landscape or portrait state as @leonardder suggests, not the width and height. What you're trying to suppress here is double reporting of landscape, so I think it's better to cache what you're trying to suppress. Otherwise, you're going to have to do a whole heap of comparisons on width and height to work out whether to suppress, since orientation changes inherently flip width and height.

derekriemer commented:
hold off on merging this. I'll fix this tonight if I'm not wiped out after not getting done with class until oh dark thirty.

Commit:
Merge branch 'master' into i4612

Commit:
Add caching

derekriemer commented:
@jcsteh go ahead

derekriemer commented:
As soon as @jcsteh reincubates this, I'm going to start a new PR to add extension points for this code, so notification of a window message can be requested. This will allow me to move the existing code somewhere else. Any suggestions on where I should move the code? maybe windowutils, ui.py (Not a huge fan), eventHandler (also kind of a no), or a custom file?

josephsl commented:
Hi, I suggest windowUtils. Thanks.

 

From: Derek Riemer [mailto:notifications@github.com] 
Sent: Sunday, September 10, 2017 10:19 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Comment <comment@noreply.github.com>
Subject: Re: [nvaccess/nvda] If you are running a laptop, NVDA will now announce changes to whether a charger is plugged in and more. #4612 (#6320)

 

As soon as @jcsteh <https://github.com/jcsteh>  reincubates this, I'm going to start a new PR to add extension points for this code, so notification of a window message can be requested. This will allow me to move the existing code somewhere else. Any suggestions on where I should move the code? maybe windowutils, ui.py (Not a huge fan), eventHandler (also kind of a no), or a custom file?

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/6320#issuecomment-328417786> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkD4J7FT6b1LAJlQsh0wovId5Mdwkks5shMKvgaJpZM4JxF3F> .

jcsteh commented on source/core.py:
```diff
@@ -286,12 +291,18 @@ def handleScreenOrientationChange(self, lParam):
 			# Resolution detection comes from an article found at https://msdn.microsoft.com/en-us/library/ms812142.aspx.
 			#The low word is the width and hiword is height.
-			if winUser.LOWORD(lParam) > winUser.HIWORD(lParam):
+			if (winUser.LOWORD(lParam) > winUser.HIWORD(lParam):
```
nit: Extraneous outer parens. I wouldn't normally comment except you don't do it anywhere else and it's unnecessarily new in this latest commit. :)

Commit:
Fix review actions, and implement screen flip detection.

derekriemer commented:
@jcsteh Sorry :-( I implemented flip detection (one motivation for this PR was to prevent users from flipping their screen without knowing). Now, if you flip the screen, portrait or landscape is announced, but if you change the resolution, it isn't. 
Testing:
* alt+ctrl+up. alt+ctrl+down. Did it speak?
* ac+right, ac+left. Did it speak?
* settings, changed resolution several times. When did it speak?
* changed orientation several times in a counterclockwise direction. It spoke for all of them.
* Changed from landscape to portrait several times. It spoke.

jcsteh commented on source/core.py:
```diff
@@ -291,18 +291,21 @@ def handleScreenOrientationChange(self, lParam):
+			width = winUser.LOWORD(lParam)
+			height = winUser.HIWORD(lParam)
+			if width > height:
+				#Check to see if the height and width are the same. If so, it's actually a screen flip, and we want to alert of those!
```
I feel like this comment could be a bit misleading, as it suggests you're checking whether they're the same (but you're actually checking whether they're different). Perhaps this:

`# If the height and width are the same, it's actually a screen flip, and we do want to alert of those!`

jcsteh commented on source/core.py:
```diff
@@ -291,18 +291,21 @@ def handleScreenOrientationChange(self, lParam):
+			height = winUser.HIWORD(lParam)
+			if width > height:
+				#Check to see if the height and width are the same. If so, it's actually a screen flip, and we want to alert of those!
+				if self.orientationCache[0] == self.ORIENTATION_LANDSCAPE and self.orientationCache[1:] != (width,height):
```
While I understand you're trying to avoid unnecessary variables here, you end up building a new tuple with `[1:]` anyway. I suggest splitting out the orientation state cache (portrait/landscape) from the orientation width height cache (x, y); e.g. `self.orientationStateCache`, `self.orientationXYCache`. That also means you only have to set the width/height cache in one place, outside the condition.

derekriemer commented:
@jcsteh

Commit:
Fix review actions

michaelDCurran commented:
@derekriemer: It seems this pr now has conflicts with master. Once these are addressed I can merge to master.

derekriemer commented:
okay, just a modification to copyright names. I just had to fix that. changes coming right up.

Commit:
Fix merge conflicts

derekriemer commented:
@michaelDCurran I believe the merge conflict is fixed. let me know if you need anything else.
